### PR TITLE
Remove mpi dependency

### PR DESCRIPTION
**Pull request by utkarshayachit**

### CODE REVIEW

**Comment by mrzv** [comment]
There was a problem with DIY and Travis. I've fixed it now. Can you rebase and resolve conflict with `tests/scripts/merge-swap-reduce.sh`?

**Comment by utkarshayachit** [comment]
@mrzv looks like all tests are passing now.

**Comment by mrzv** [comment]
@utkarshayachit Thanks. I have some questions/comments that I'll leave inline. Hopefully, later today.

File: include/diy/mpi/io.hpp [hunk 1]
```diff
@@ -70,12 +67,62 @@ namespace io
...
 }
 
+diy::mpi::io::file::
+file(const communicator& comm, const std::string& filename, int mode)
+: comm_(comm)
+{
+#ifndef DIY_NO_MPI
+  MPI_File_open(comm, const_cast<char*>(filename.c_str()), mode, MPI_INFO_NULL, &fh);
+#else
+  (void) filename; (void) mode;
```
Comment (mrzv) [inline include/diy/mpi/io.hpp]: Why no `DIY_UNSUPPORTED_MPI_CALL` here?

I'm not sure I agree that we can't support `diy::mpi::io::file` in the `DIY_NO_MPI` setting. Most of the operations (unlike the subarrays in BOV) are fairly straightforward to implement. I guess it's Ok for now, but we should reconsider this in the future.

My first comment stands: why not throw right from the constructor?

File: include/diy/mpi/communicator.hpp [hunk 2]
```diff
@@ -58,15 +60,54 @@ namespace mpi
...
 {
+  (void) source;
+  (void) tag;
+#ifndef DIY_NO_MPI
   status s;
   int flag;
   MPI_Iprobe(source, tag, comm_, &flag, &s.s);
   if (flag)
     return s;
+#endif
```
Comment (mrzv) [inline include/diy/mpi/communicator.hpp]: This seems a bit treacherous to me. Maybe both `probe` and `iprobe` should throw via `DIY_UNSUPPORTED_MPI_CALL`? I can't think of when they'd be used correctly in the no-mpi mode, and especially when returning empty status would be meaningful.

File: include/diy/mpi/io.hpp (hunk 1 above)
Comment (utkarshayachit) [reply include/diy/mpi/io.hpp]: good catch..just oversight..I added `DIY_UNSUPPORTED_MPI_CALL` after a few steps in when I was updating the code.

File: include/diy/mpi/communicator.hpp (hunk 2 above)
Comment (utkarshayachit) [reply include/diy/mpi/communicator.hpp]: sounds good.

File: include/diy/mpi/collectives.hpp [hunk 3]
```diff
@@ -32,33 +35,42 @@ namespace mpi
...
 
     static request ibroadcast(const communicator& comm, T& x, int root)
     {
+#ifndef DIY_NO_MPI
       request r;
       MPI_Ibcast(Datatype::address(x),
                  Datatype::count(x),
                  Datatype::datatype(), root, comm, &r.r);
       return r;
+#endif
```
Comment (mrzv) [inline include/diy/mpi/collectives.hpp]: You need to return something here. Perhaps, only `MPI_Ibcast` needs to be wrapped in the guards?

Comment (mrzv) [reply include/diy/mpi/collectives.hpp]: Although, the entire request system in the no-mpi case is a bit shaky. Maybe it's best to add `DIY_UNSUPPORTED_MPI_CALL` to trigger an error. Then we can deal with it, with a specific use case in mind.

Comment (utkarshayachit) [reply include/diy/mpi/collectives.hpp]: doh, yup.

**Comment by mrzv** [comment]
@utkarshayachit Overall, this looks great. Take a look at the few inline comments I left, but it's mostly ready to be merged.

**Comment by utkarshayachit** [comment]
looks like the iprobe exposed some issues. I'll track that down tomorrow.

File: include/diy/mpi/communicator.hpp (hunk 2 above)
Comment (utkarshayachit) [reply include/diy/mpi/communicator.hpp]: @mrzv this is not working as expected. iprobe is used in `master::comm_exchange`. On second thought, return a status indicating nothing available here when not built with MPI seems reasonable to me, as that's indeed true. What do you think? Alternatively, I can `ifdef` out the iprobe call in `master::comm_exchange`.

**Comment by utkarshayachit** [comment]
@mrzv iprobe change is not working. iprobe is used in master::comm_exchange. On second thought, return a status indicating nothing available here when not built with MPI seems reasonable to me, as that's indeed true. What do you think? Alternatively, I can ifdef out the iprobe call in master::comm_exchange.

**Comment by mrzv** [comment]
Yes, I see what you mean. And I agree: `iprobe` should return that nothing is available. `probe` on the other hand should be unsupported.

**Comment by utkarshayachit** [comment]
`probe` is already flagged as unsupported and should throw the exception. I've updated the MR accorindly.

**Comment by mrzv** [comment]
Great. Thanks, Utkarsh. I'll scan through the changes again, but assuming nothing else jumps out at me, I'll merge this in.

**Comment by mrzv** [comment]
Merged. Thanks again, Utkarsh.

**Pull request merged**